p11-kit/generate-keypair.c

```diff
@@ -0,0 +1,412 @@
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * Author: Zoltan Fridrich <zfridric@redhat.com>
+ */
+
+#include "config.h"
+
+#include "attrs.h"
+#include "buffer.h"
+#include "compat.h"
+#include "debug.h"
+#include "iter.h"
+#include "message.h"
+#include "mock.h"
+#include "tool.h"
+
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef ENABLE_NLS
+#include <libintl.h>
+#define _(x) dgettext(PACKAGE_NAME, x)
+#else
+#define _(x) (x)
+#endif
+
+#define cleanup_if_fail(X, S)          \
+	do { if (X) { } else {         \
+		p11_message (_(S));    \
+		goto cleanup;          \
+	} } while (false)
+
+#define cleanup_if_fail_rv(X, S, R)    \
+	do { if (X) { } else {         \
+		p11_message (_(S), R); \
+		goto cleanup;          \
+	} } while (false)
+
+const CK_BBOOL TVAL = CK_TRUE;
+const CK_BBOOL FVAL = CK_FALSE;
+
+int
+p11_kit_generate_keypair (int argc,
+			  char *argv[]);
+
+static bool
+get_mechanism (CK_MECHANISM *m,
+	       const char *type)
+{
+	if (p11_ascii_strcaseeq (type, "mock")) {
+		m->mechanism = CKM_MOCK_GENERATE;
+		m->pParameter = "generate";
+		m->ulParameterLen = 9;
+		return true;
+	}
+
+	if (p11_ascii_strcaseeq (type, "rsa"))
+		m->mechanism = CKM_RSA_PKCS_KEY_PAIR_GEN;
+	else if (p11_ascii_strcaseeq (type, "ecdsa"))
+		m->mechanism = CKM_ECDSA_KEY_PAIR_GEN;
+	else if (p11_ascii_strcaseeq (type, "ed25519"))
+		m->mechanism = CKM_EC_EDWARDS_KEY_PAIR_GEN;
+	else
+		return false;
+
+	m->pParameter = NULL_PTR;
+	m->ulParameterLen = 0;
+
+	return true;
+}
+
+static bool
+get_ec_params (const char *curve,
+	       p11_buffer *ec_params)
+{
+	bool ok = true;
+	const char *data = NULL;
+	size_t len = 0;
+
+	if (p11_ascii_strcaseeq (curve, "secp256r1")) {
+		data = "\x06\x08\x2A\x86\x48\xCE\x3D\x03\x01\x07";
+		len = 10;
+	} else if (p11_ascii_strcaseeq (curve, "secp384r1")) {
+		data = "\x06\x05\x2B\x81\x04\x00\x22";
+		len = 7;
+	} else if (p11_ascii_strcaseeq (curve, "secp521r1")) {
+		data = "\x06\x05\x2B\x81\x04\x00\x23";
+		len = 7;
+	} else
+		ok = false;
+
+	p11_buffer_init_full (ec_params, (void *)data, len, 0, NULL, NULL);
+	return ok;
+}
+
+static bool
+check_args (CK_MECHANISM_TYPE type,
+	    CK_ULONG bits,
+	    const p11_buffer *ec_params)
+{
+	if (type == CKA_INVALID) {
+		p11_message (_("no type specified"));
+		return false;
+	}
+
+	switch (type) {
+	case CKM_MOCK_GENERATE:
+		break;
+	case CKM_RSA_PKCS_KEY_PAIR_GEN:
+		if (bits == 0) {
+			p11_message (_("no bits specified"));
+			return false;
+		}
+		break;
+	case CKM_ECDSA_KEY_PAIR_GEN:
+	case CKM_EC_EDWARDS_KEY_PAIR_GEN:
+		if (ec_params->data == NULL) {
+			p11_message (_("no curve specified"));
+			return false;
+		}
+		break;
+	default:
+		p11_message (_("unkwnown mechanism type in %s"), __func__);
+		return false;
+	}
+
+	if (bits != 0 && ec_params->data != NULL) {
+		p11_message (_("both bits and curve cannot be specified at once"));
```

Comment:
Since this message is translated, making this a little more generic would reduce translators burden: `p11_message (_("both %s and %s cannot be specified at once"), "--bits", "--curves")`, so no new message will need to be translated even after we add another options mutually exclusive.